go-controller/pkg/allocator/udn/ips.go

```diff
@@ -0,0 +1,76 @@
+package udn
+
+import (
+	"fmt"
+	"net"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/config"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/types"
+	utilnet "k8s.io/utils/net"
+)
+
+const (
+	masqueradeIPv4IDName = "v4-masquerade-ips"
+	masqueradeIPv6IDName = "v6-masquerade-ips"
+)
+
+type MasqueradeIPs struct {
+	Shared net.IP
+	Local  net.IP
+}
+
+// AllocateV4MasqueradeIPs will return local and shared gateway masquerade IPv4 addresses calculated from
+// the networkID argument
+func AllocateV4MasqueradeIPs(networkID int) (*MasqueradeIPs, error) {
+	return allocateMasqueradeIPs(masqueradeIPv4IDName, config.Gateway.V4MasqueradeSubnet, networkID)
+}
+
+// AllocateV6MasqueradeIPs will return local and shared gateway masquerade IPv6 addresses calculated from
+// the networkID argument
+func AllocateV6MasqueradeIPs(networkID int) (*MasqueradeIPs, error) {
+	return allocateMasqueradeIPs(masqueradeIPv6IDName, config.Gateway.V6MasqueradeSubnet, networkID)
+}
+
+func allocateMasqueradeIPs(idName string, masqueradeSubnet string, networkID int) (*MasqueradeIPs, error) {
+	if networkID < 1 {
+		return nil, fmt.Errorf("invalid argument: network ID should be bigger that 0")
+	}
+	subnetIP, subnetCIDR, err := net.ParseCIDR(masqueradeSubnet)
+	if err != nil {
+		return nil, err
+	}
+	// Let's illustrate the expected IPs for networkID 1 and 2
+	// with UserDefinedNetworkMasqueradeIPBase=10 and subnet=169.254.169.0/16
+	// networkID=1
+	// 	Shared: 169.254.169.11
+	// 	Local: 169.254.169.12
+	// networkID=2
+	//  Shared: 169.254.169.13
+	//  Local: 169.254.169.14
+	//
+	// So the calculations for networkID=2 is done by first position the
+	// previousNetworkLastIPOffset to 12 and then increment from there
+	// until numberOfIPs which is 2 that is 13 and 14, those values are
+	// added to the subnet (169.254.169.0) with AddIPOffset to convert the offset to the
+	// masquerade IPs 169.254.169.13 and 169.254.169.14
+	numberOfIPs := 2
+	previousNetworkID := networkID - 1
```

Comment:
and just to be sure the networkIDs are serially getting allocated correct? can there be a gap value or something where a number never got allocated? meaning networkID is valid given we will pass it from netInfo, using networkID-1 as the previous network's ID is that a guaranteed scenario always?

Reply:
> and just to be sure the networkIDs are serially getting allocated correct? 
We don't care if they are serial or not, we just care about the stuff starting from our network-id.

> can there be a gap value or something where a number never got allocated? 
Even if there is a gap it does not affect the fact that we are going to start from the networkID

> meaning networkID is valid given we will pass it from netInfo, using networkID-1 as the previous network's ID is that a guaranteed scenario always?
We know for sure that our networks start from 1 so the lower limit of network-1 is going to be 0 so all is good.